### PR TITLE
Detect Slashable Attestations in Optimized Slasher

### PR DESCRIPTION
**Pull request by rauljordan**
This PR detects slashable attestations in an optimized slasher implementation from feature/slasher. Diff is mostly tests

### CODE REVIEW

File: beacon-chain/slasher/detect_attestations.go [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+	// TODO(#8331): Consider using goroutines and wait groups here.
+	groupedAtts := s.groupByValidatorChunkIndex(atts)
+	for validatorChunkIdx, batch := range groupedAtts {
+		attSlashings, err := s.detectAllAttesterSlashings(ctx, &chunkUpdateArgs{
+			validatorChunkIndex: validatorChunkIdx,
+			currentEpoch:        currentEpoch,
+		}, batch)
+
+		slashings = append(slashings, attSlashings...)
+		if err != nil {
```
Comment (terencechain) [inline beacon-chain/slasher/detect_attestations.go:33]: this err is at wrong place

Comment (prestonvanloon) [reply beacon-chain/slasher/detect_attestations.go:32]: Seems good to me. Whats wrong with it? @terencechain

Comment (terencechain) [reply beacon-chain/slasher/detect_attestations.go:32]: the err is after the `append`, it should be before the `append`

Comment (terencechain) [reply beacon-chain/slasher/detect_attestations.go:32]: It's fixed in https://github.com/prysmaticlabs/prysm/pull/9694/commits/6eeb088a69d4307c25966610e311ba9d823ac84b

File: beacon-chain/slasher/detect_attestations.go [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"could not update max attestation spans for validator chunk index %d",
+			args.validatorChunkIndex,
+		)
+	}
+
+	// Consolidate all slashings into a slice.
+	slashings := make([]*ethpb.AttesterSlashing, 0)
```
Comment (prestonvanloon) [inline beacon-chain/slasher/detect_attestations.go:97]:
```suggestion
	slashings := make([]*ethpb.AttesterSlashing, 0, len(doubleVoteSlashings)+len(surroundingSlashings)+len(surroundedSlashings))
```

**Review by prestonvanloon — APPROVED** [review]
One suggestion, otherwise LGTM

**Comment by rauljordan** [comment]
Applied, ready @prestonvanloon

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**